File: core/store/src/main/java/org/locationtech/geowave/core/store/api/QueryConstraintsFactory.java
/**
 * Copyright (c) 2013-2022 Contributors to the Eclipse Foundation
 *
 * <p> See the NOTICE file distributed with this work for additional information regarding copyright
 * ownership. All rights reserved. This program and the accompanying materials are made available
 * under the terms of the Apache License, Version 2.0 which accompanies this distribution and is
 * available at http://www.apache.org/licenses/LICENSE-2.0.txt
 */
package org.locationtech.geowave.core.store.api;

import org.locationtech.geowave.core.index.MultiDimensionalCoordinateRangesArray;
import org.locationtech.geowave.core.index.NumericIndexStrategy;
import org.locationtech.geowave.core.index.persist.Persistable;
import org.locationtech.geowave.core.store.index.CustomIndex;
import org.locationtech.geowave.core.store.query.constraints.Constraints;
import org.locationtech.geowave.core.store.query.constraints.QueryConstraints;
import org.locationtech.geowave.core.store.query.filter.BasicQueryFilter.BasicQueryCompareOperation;

/** This is a simple mechanism to create existing supported query constraints. */
public interface QueryConstraintsFactory {
  /**
   * constrain a query by data IDs
   *
   * @param dataIds the data IDs to constrain to
   * @return the constraints
   */
  QueryConstraints dataIds(final byte[]... dataIds);

  /**
   * constrain a query using a range of data IDs, assuming big endian ordering
   *
   * @param startDataIdInclusive the start of data ID range (inclusive)
   * @param endDataIdInclusive the end of data ID range (inclusive)
   * @return the constraints
   */
  QueryConstraints dataIdsByRange(
      final byte[] startDataIdInclusive,
      final byte[] endDataIdInclusive);

  /**
   * constrain a query using a range of data IDs, assuming big endian ordering
   *
   * RocksDB and HBase are currently the only two that will support this, but allows for reverse
   * iteration from "end" to "start" data ID
   *
   * All other datastores will throw an UnsupportedOperationException and the forward scan should be
   * preferred for those datastores
   *
   * @param startDataIdInclusive the start of data ID range (inclusive)
   * @param endDataIdInclusive the end of data ID range (inclusive)
   * @return the constraints
   */
  QueryConstraints dataIdsByRangeReverse(
      final byte[] startDataIdInclusive,
      final byte[] endDataIdInclusive);

  /**
   * constrain a query by prefix
   *
   * @param partitionKey the prefix
   * @param sortKeyPrefix the sort prefix
   * @return the constraints
   */
  QueryConstraints prefix(final byte[] partitionKey, final byte[] sortKeyPrefix);

  /**
   * constrain by coordinate ranges
   *
   * @param indexStrategy the index strategy
   * @param coordinateRanges the coordinate ranges
   * @return the constraints
   */
  QueryConstraints coordinateRanges(
      final NumericIndexStrategy indexStrategy,
      final MultiDimensionalCoordinateRangesArray[] coordinateRanges);

  /**
   * constrain generally by constraints
   *
   * @param constraints the constraints
   * @return the query constraints
   */
  QueryConstraints constraints(final Constraints constraints);

  /**
   * constrain generally by constraints with a compare operation
   *
   * @param constraints the constraints
   * @param compareOp the relationship to use for comparison
   * @return the query constraints
   */
  QueryConstraints constraints(
      final Constraints constraints,
      final BasicQueryCompareOperation compareOp);

  /**
   * constrain using a custom persistable object NOTE: this only applies to an index that is a
   * {@link CustomIndex} and the instance of these custom constraints must match the generic of the
   * custom index's strategy
   *
   * @param customConstraints the instance of custom constraints
   * @return the query constraints
   */
  QueryConstraints customConstraints(final Persistable customConstraints);

  /**
   * no query constraints, meaning wide open query (this is the default)
   *
   * @return the query constraints
   */
  QueryConstraints noConstraints();
}

File: python/src/main/python/pygw/query/query_constraints_factory.py
#
# Copyright (c) 2013-2020 Contributors to the Eclipse Foundation

#
# See the NOTICE file distributed with this work for additional information regarding copyright
# ownership. All rights reserved. This program and the accompanying materials are made available
# under the terms of the Apache License, Version 2.0 which accompanies this distribution and is
# available at http://www.apache.org/licenses/LICENSE-2.0.txt
#===============================================================================================

from py4j.java_gateway import JavaClass

from pygw.base import GeoWaveObject
from pygw.base.type_conversions import PrimitiveByteArrayType
from pygw.config import geowave_pkg
from pygw.config import java_gateway

from .query_constraints import QueryConstraints

_pbat = PrimitiveByteArrayType()

class QueryConstraintsFactory(GeoWaveObject):
    """
    Base factory for constructing general query constraints to be used by the
    query builder.  Do not use this factory directly, instead get an instance of
    the factory from the `pygw.query.query_builder.QueryBuilder.constraints_factory` method of the query builder.
    """

    def data_ids(self, data_ids):
        """
        Constrain a query by data IDs.

        Args:
            data_ids (list of bytes): The data IDs to constrain by.
        Returns:
            A `pygw.query.query_constraints.QueryConstraints` with the given data ids.
        """
        byte_array_class = JavaClass("[B", java_gateway._gateway_client)
        j_data_ids = java_gateway.new_array(byte_array_class, len(data_ids))
        for idx, data_id in enumerate(data_ids):
            j_data_ids[idx] = _pbat.to_java(data_id)
        j_qc = self._java_ref.dataIds(j_data_ids)
        return QueryConstraints(j_qc)

    def data_ids_by_range(self, start_data_id_inclusive, end_data_id_inclusive):
        """
        Constrain a query using a range of data IDs, assuming big endian ordering.

        Args:
            start_data_id_inclusive (bytes): The start of data ID range (inclusive).
            end_data_id_inclusive (bytes): The end of data ID range (inclusive).
        Returns:
            A `pygw.query.query_constraints.QueryConstraints` with the given data ID range.
        """
        j_qc = self._java_ref.dataIdsByRange(_pbat.to_java(start_data_id_inclusive),
                                             _pbat.to_java(end_data_id_inclusive))
        return QueryConstraints(j_qc)

    def prefix(self, partition_key, sort_key_prefix):
        """
        Constrain a query by prefix.

        Args:
            partition_key (bytes): The partition key to constrain by.
            sort_key_prefix (bytes): The sort key prefix to constrain by.
        Returns:
            A `pygw.query.query_constraints.QueryConstraints` with the given prefix.
        """
        j_qc = self._java_ref.prefix(_pbat.to_java(partition_key),
                                     _pbat.to_java(sort_key_prefix))
        return QueryConstraints(j_qc)

    def coordinate_ranges(self, index_strategy, coordinate_ranges):
        # TODO: support Python variants of NumericIndexStrategy and MultiDimensionalCoordinateRangesArray
        raise NotImplementedError

    def constraints(self, constraints, compare_op=None):
        # TODO: support Python variants of Constraints
        raise NotImplementedError

    def no_constraints(self):
        """
        Use no query constraints, meaning wide open query (this is the default).

        Returns:
            A `pygw.query.query_constraints.QueryConstraints` with no constraints.
        """
        return QueryConstraints(self._java_ref.noConstraints())
